numba/tests/test_array_exprs.py

```diff
@@ -627,5 +645,127 @@ def do_call(x, y):
         self.assertTrue(isinstance(oty.type.dtype, types.Float))
 
 
+class TestExternalTypes(MemoryLeakMixin, unittest.TestCase):
+    """ Tests RewriteArrayExprs with external (user defined) types,
+    see #5157"""
+
+    source_lines = textwrap.dedent("""
+        from numba.core import types
+
+        class FooType(types.Type):
+            def __init__(self):
+                super(FooType, self).__init__(name='Foo')
+        """)
+
+    # copied from test_types
+    @contextlib.contextmanager
+    def create_temp_module(self, source_lines=None, **jit_options):
+        # Use try/finally so cleanup happens even when an exception is raised
+        try:
+            if source_lines is None:
+                source_lines = self.source_lines
+            tempdir = temp_directory('test_extension_type')
+            # Generate random module name
+            temp_module_name = 'test_extension_type_{}'.format(
+                str(uuid.uuid4()).replace('-', '_'))
+            temp_module_path = os.path.join(tempdir, temp_module_name + '.py')
+
+            with open(temp_module_path, 'w') as f:
+                lines = source_lines.format(jit_options=jit_options)
+                f.write(lines)
+            # Add test_module to sys.path so it can be imported
+            sys.path.insert(0, tempdir)
+            test_module = importlib.import_module(temp_module_name)
+            yield test_module
+        finally:
+            sys.modules.pop(temp_module_name, None)
+            sys.path.remove(tempdir)
+            shutil.rmtree(tempdir)
+
```

Comment:
It'd be nice to refactor this into `numba.tests.support`, but perhaps not for this PR.

Reply:
https://github.com/numba/numba/issues/6778 tracks